bin/write_viral_gff.py

```diff
@@ -21,136 +20,213 @@ def get_ena_contig_mapping(ena_contig_file):
-                #   set to unclassified if all levels are empty
-                if joined_lineage == ";;;":
-                    taxonomy_dict[row["contig_ID"]] = "unclassified"
+        with open(tfile,'r') as input_table:
+            next(input_table)
+            for line in input_table:
+                line_l=line.rstrip().split('\t')
+                Contig=line_l.pop(0)
+
+                if '|phage-circular' in Contig:
+                    Contig=Contig.replace('|phage-circular','')
+                elif '|prophage-' in Contig:
+                    desc='prophage'
                 else:
-                    taxonomy_dict[row["contig_ID"]] = joined_lineage
-
-    with open(output_filename, "w") as metadata:
-        metadata.write(headers)
-        for contig in virify_contigs:
-            #   change contig name if ena mapping required
-            if ena_mapping:
-                contig_name = f"{ena_mapping[contig]}-{contig}"
+                    Contig=Contig.replace('|','')
+
+                lineage=[]
+                for element in line_l:
+                    if len(element)>0:
+                        try:
+                            float(element)
+                        except:
+                            lineage.append(element)
+                if len(lineage)>0:
+                    taxo_string='%3B'.join(lineage)
+                else:
+                    taxo_string='unclassified'
+
+                taxonomy_dict[Contig]=taxo_string
+
+    # Saving the original contig length
+    for record in SeqIO.parse(assembly_file, "fasta"):
+        contig_id=str(record.id)
+        seq_len=len(str(record.seq))
+        contigs_len_dict[contig_id]=str(seq_len)
+
+    contig_protID=annot_dicts[0]
+    prot_data=annot_dicts[1]
+    contig_data=annot_dicts[2]
+    # Writing the gff file
+    with open(output_filename, "w") as gff:
+        gff.write("##gff-version 3\n")
+        contig_protID=annot_dicts[0]
+        prot_data=annot_dicts[1]
+        contig_data=annot_dicts[2]
+
+        # Writing the gff header
+        printed_contig=[]
+        for element in contig_protID.keys():
+            if 'prophage' in element:
+                clean_id=element.split('|')[0]
+            else:
+                clean_id=element
+
+            if clean_id not in printed_contig:
+                printed_contig.append(clean_id)
+                element_len=contigs_len_dict[clean_id]
+                gff.write('##sequence-region '+clean_id+' 1 '+element_len+'\n')
+
+        # Writing the mobile genetic elements coordinates and attributes
+        phage_counter=0
+        phage_ids={}
+        for element in contig_protID.keys():
+            phage_counter+=1
+            element_id='phage_'+str(phage_counter)
+            phage_ids[element]=element_id
+            if 'prophage' in element:
+                seq_type='prophage'
+                seqid=element.split('|')[0]
+                start=element.split('|')[1].split('-')[1].split(':')[0]
```

Comment:
These split calls are repeated in several places? Do you think we can write a method that returns the values instead if repeating the split calls?